### ISSUE TITLE
"PyTorch Fail" in Dependencies check of SCP QGIS

### ISSUE THREAD

**Issue by chinkgheirk** [SHI-Labs/NATTEN#251]
Hi Everyone,
I have just started shifting from ARCGIS to QGIS and got the following issue while trying to execute the SCP plugin. I have installed the Torch but still got the issue. Kindly help me out. below is the debug log file.

INFO|2025-08-12T20:14:01.676|system_tools|get_system_info|38|version:0.5.1; system: Darwin; 64bit: True; n_processes: 2; ram: 2048; temp.dir: /private/var/folders/57/mx86lqjd62j_tvwlgs7pkvcm0000gn/T/semiautomatic_classification/remotior_sensus_0812eeh6dt3c
WARNING|2025-08-12T20:14:01.676|session|_check_dependencies|490|No module named 'torch'
INFO|2025-08-12T20:14:01.819|bandset_catalog|create|745|start
INFO|2025-08-12T20:14:01.820|bandset_catalog|create|751|end
DEBUG|2025-08-12T20:14:01.821|bandset_catalog|get_band_attributes|344|attribute: name
DEBUG|2025-08-12T20:14:01.821|bandset_catalog|_empty_bandset|2582|empty bandset
INFO|2025-08-12T20:14:01.822|semiautomaticclassificationplugin|__init__|411|SCP Version 8.5.0 - Infinity; QGIS v. 34201; L: en; OS: Darwin; python: /Applications/QGIS.app/Contents/MacOS/bin/python3; remotior sensus: 0.5.1; environment: None
INFO|2025-08-12T20:14:01.879|session|set|442|n_processes: 2; ram: 2048; temp.dir: /private/var/folders/57/mx86lqjd62j_tvwlgs7pkvcm0000gn/T/semiautomatic_classification/remotior_sensus_0812eeh6dt3c
DEBUG|2025-08-12T20:14:01.882|semiautomaticclassificationplugin|initGui|556|logger: 2
DEBUG|2025-08-12T20:14:01.920|semiautomaticclassificationplugin|initGui|756|init GUI
DEBUG|2025-08-12T20:14:01.922|semiautomaticclassificationplugin|connect_gui|1891|GUI connected
DEBUG|2025-08-12T20:14:03.871|semiautomaticclassificationplugin|new_project_loaded|791|new_project_loaded
DEBUG|2025-08-12T20:14:03.871|semiautomaticclassificationplugin|reset_scp|1896|reset_scp
INFO|2025-08-12T20:14:03.874|bandset_catalog|create|745|start
INFO|2025-08-12T20:14:03.874|bandset_catalog|create|751|end
DEBUG|2025-08-12T20:14:03.874|bandset_catalog|get_band_attributes|344|attribute: name
DEBUG|2025-08-12T20:14:03.874|bandset_catalog|_empty_bandset|2582|empty bandset
DEBUG|2025-08-12T20:14:03.874|bandset_tab|add_band_set_tab|188|add_band_set_tab position: None
DEBUG|2025-08-12T20:14:03.877|bandset_tab|add_band_set_tab|286|bandset_count: 1; table count: 1; current index: 0
DEBUG|2025-08-12T20:14:03.877|scp_dock|clear_tree|573|clear_tree
DEBUG|2025-08-12T20:14:03.903|utils|read_project_variables|765|project_registry: {'scp_roi_index_calculation_check': 0, 'scp_custom_index_calculation': '', 'scp_rgb_list': ['-', '3-2-1', '4-3-2', '7-3-2'], 'scp_signature_calculation_check': 2, 'scp_rapid_roi_check': 2, 'scp_save_training_input_check': 2, 'scp_roi_main_band': 1, 'scp_roi_max_width': 100, 'scp_roi_min_size': 60, 'scp_roi_range_radius': 0.01, 'scp_roi_custom_expression': '', 'scp_reg_preview_size': 200, 'scp_roi_class_id': 1, 'scp_roi_class_name': 'Class 1', 'scp_roi_macroclass_id': 1, 'scp_roi_macroclass_name': 'Macroclass 1', 'scp_training_input_path': '', 'scp_active_bandset_number': 1, 'scp_training_bandset_number': 1, 'scp_download_product_table': '', 'scp_bandset_count': 1}
DEBUG|2025-08-12T20:14:03.903|utils|set_rgb_color_composite|647|set_rgb_color_composite rgb: -
DEBUG|2025-08-12T20:14:03.903|utils|create_rgb_color_composite|688|create_rgb_color_composite bandset: 1
DEBUG|2025-08-12T20:14:03.904|bandset_catalog|get_bandset|1965|bandset_number: 1
DEBUG|2025-08-12T20:14:03.904|utils|create_rgb_color_composite|702|empty bandset
DEBUG|2025-08-12T20:14:03.905|utils|read_project_variables|831|bandset restore: 1
DEBUG|2025-08-12T20:14:03.905|bandset_tab|add_band_set_tab|188|add_band_set_tab position: 1
DEBUG|2025-08-12T20:14:03.905|bandset_tab|add_band_set_tab|201|existing bandset tab
DEBUG|2025-08-12T20:14:03.905|bandset_catalog|import_bandset_from_xml|2612|import bandset from xml
DEBUG|2025-08-12T20:14:03.905|bandset_catalog|get_bandset|1965|bandset_number: 1
DEBUG|2025-08-12T20:14:03.905|bandset_catalog|import_as_xml|619|import bandset: /private/var/folders/57/mx86lqjd62j_tvwlgs7pkvcm0000gn/T/semiautomatic_classification/remotior_sensus_0812eeh6dt3c/t20250812_201403905165_9420.xml
DEBUG|2025-08-12T20:14:03.905|bandset_tab|band_set_to_table|902|band_set_to_table bandset_number: 1
DEBUG|2025-08-12T20:14:03.905|bandset_catalog|get_bandset_by_number|1690|number: 1
DEBUG|2025-08-12T20:14:03.905|bandset_catalog|get_bandset|1965|bandset_number: 1
DEBUG|2025-08-12T20:14:03.905|bandset_tab|band_set_to_table|915|band_count: 0
DEBUG|2025-08-12T20:14:03.906|bandset_catalog|get_band_attributes|344|attribute: name
DEBUG|2025-08-12T20:14:03.906|bandset_catalog|get_bandset_catalog_attributes|1907|number: 1; attribute: date
DEBUG|2025-08-12T20:14:03.906|bandset_catalog|get_bandset_catalog_attributes|1919|result: NaT
DEBUG|2025-08-12T20:14:03.906|bandset_catalog|get_bandset_catalog_attributes|1907|number: 1; attribute: root_directory
DEBUG|2025-08-12T20:14:03.906|bandset_catalog|get_bandset_catalog_attributes|1919|result: None
DEBUG|2025-08-12T20:14:03.906|download_products_tab|open_download_table|516|open_download_table: None
DEBUG|2025-08-12T20:14:12.088|multiprocess_manager|run|146|multiprocess: <function raster_unique_values_with_sum at 0x1cace51f0>
DEBUG|2025-08-12T20:14:12.088|multiprocess_manager|_calculate_block_size|2510|start
DEBUG|2025-08-12T20:14:12.120|raster_vector|raster_info|252|/Users/chingkheirk/Library/Application Support/QGIS/QGIS3/profiles/default/python/plugins/SemiAutomaticClassificationPlugin/debug/raster.tif :[(304545.0, 30.0, 0.0, 4624365.0, 0.0, -30.0), 'metre', [8, 8], -99999.0, 1, [8, 8], [None, None], 'Float32', 'PROJCS["WGS 84 / UTM zone 33N",GEOGCS["WGS 84",DATUM["WGS_1984",SPHEROID["WGS 84",6378137,298.257223563,AUTHORITY["EPSG","7030"]],AUTHORITY["EPSG","6326"]],PRIMEM["Greenwich",0,AUTHORITY["EPSG","8901"]],UNIT["degree",0.0174532925199433,AUTHORITY["EPSG","9122"]],AUTHORITY["EPSG","4326"]],PROJECTION["Transverse_Mercator"],PARAMETER["latitude_of_origin",0],PARAMETER["central_meridian",15],PARAMETER["scale_factor",0.9996],PARAMETER["false_easting",500000],PARAMETER["false_northing",0],UNIT["metre",1,AUTHORITY["EPSG","9001"]],AXIS["Easting",EAST],AXIS["Northing",NORTH],AUTHORITY["EPSG","32633"]]']
DEBUG|2025-08-12T20:14:12.120|multiprocess_manager|_calculate_block_size|2561|end; r_x: 8; r_y: 8; block_size_x: 8; block_size_y: 5; list_range_x: [0]; list_range_y: [0, 5]; tot_blocks: 2; number_of_bands: 1
DEBUG|2025-08-12T20:14:12.120|multiprocess_manager|_compute_raster_pieces|2886|len(pieces): 2
DEBUG|2025-08-12T20:14:13.127|multiprocess_manager|run|321|DEBUG_p0|2025-08-12T20:14:12.805|processor|function_initiator|122|start
DEBUG_p0|2025-08-12T20:14:12.805|processor|function_initiator|123|keep_output_array: False; keep_output_argument: True; any_nodata_mask: True; input_nodata_as_value: None
DEBUG_p0|2025-08-12T20:14:12.805|processor|function_initiator|142|raster_list: ['/Users/chingkheirk/Library/Application Support/QGIS/QGIS3/profiles/default/python/plugins/SemiAutomaticClassificationPlugin/debug/raster.tif']; x_size_piece: 8; y_size_piece: 5:
DEBUG_p0|2025-08-12T20:14:12.805|processor|function_initiator|153|process raster: /Users/chingkheirk/Library/Application Support/QGIS/QGIS3/profiles/default/python/plugins/SemiAutomaticClassificationPlugin/debug/raster.tif; calculation_datatype: <class 'numpy.float32'>
DEBUG_p0|2025-08-12T20:14:12.812|raster_vector|raster_info|252|/Users/chingkheirk/Library/Application Support/QGIS/QGIS3/profiles/default/python/plugins/SemiAutomaticClassificationPlugin/debug/raster.tif :[(304545.0, 30.0, 0.0, 4624365.0, 0.0, -30.0), 'metre', [8, 8], -99999.0, 1, [8, 8], [None, None], 'Float32', 'PROJCS["WGS 84 / UTM zone 33N",GEOGCS["WGS 84",DATUM["WGS_1984",SPHEROID["WGS 84",6378137,298.257223563,AUTHORITY["EPSG","7030"]],AUTHORITY["EPSG","6326"]],PRIMEM["Greenwich",0],UNIT["degree",0.0174532925199433,AUTHORITY["EPSG","9122"]],AUTHORITY["EPSG","4326"]],PROJECTION["Transverse_Mercator"],PARAMETER["latitude_of_origin",0],PARAMETER["central_meridian",15],PARAMETER["scale_factor",0.9996],PARAMETER["false_easting",500000],PARAMETER["false_northing",0],UNIT["metre",1,AUTHORITY["EPSG","9001"]],AXIS["Easting",EAST],AXIS["Northing",NORTH],AUTHORITY["EPSG","32633"]]']
DEBUG_p0|2025-08-12T20:14:12.812|processor|function_initiator|167|r_gt: (304545.0, 30.0, 0.0, 4624365.0, 0.0, -30.0)
DEBUG_p0|2025-08-12T20:14:12.812|processor|function_initiator|214|sections: [<remotior_sensus.core.processor.RasterSection object at 0x10bbd47f0>]
DEBUG_p0|2025-08-12T20:14:12.812|processor|function_initiator|425|_input_array.shape: (5, 8, 1)
DEBUG_p0|2025-08-12T20:14:12.812|processor|function_initiator|430|raster: /Users/chingkheirk/Library/Application Support/QGIS/QGIS3/profiles/default/python/plugins/SemiAutomaticClassificationPlugin/debug/raster.tif; band_number: 1
DEBUG_p0|2025-08-12T20:14:12.812|raster_vector|band_read_array_block|521|pixel_start_column: 0; pixel_start_row: 0; block_columns: 8; block_row: 5
DEBUG_p0|2025-08-12T20:14:12.813|processor|function_initiator|486|calculation_datatype: <class 'numpy.float32'>
DEBUG_p0|2025-08-12T20:14:12.813|processor|function_initiator|497|_input_array.shape: (5, 8, 1)
DEBUG_p0|2025-08-12T20:14:12.813|processor_functions|raster_unique_values_with_sum|1027|start
DEBUG_p0|2025-08-12T20:14:12.815|processor_functions|raster_unique_values_with_sum|1041|end
DEBUG_p0|2025-08-12T20:14:12.815|processor|function_initiator|564|output_raster_list: [None]
DEBUG_p0|2025-08-12T20:14:12.827|processor|function_initiator|683|end

DEBUG|2025-08-12T20:14:13.128|multiprocess_manager|run|321|DEBUG_p1|2025-08-12T20:14:12.805|processor|function_initiator|122|start
DEBUG_p1|2025-08-12T20:14:12.805|processor|function_initiator|123|keep_output_array: False; keep_output_argument: True; any_nodata_mask: True; input_nodata_as_value: None
DEBUG_p1|2025-08-12T20:14:12.805|processor|function_initiator|142|raster_list: ['/Users/chingkheirk/Library/Application Support/QGIS/QGIS3/profiles/default/python/plugins/SemiAutomaticClassificationPlugin/debug/raster.tif']; x_size_piece: 8; y_size_piece: 3:
DEBUG_p1|2025-08-12T20:14:12.805|processor|function_initiator|153|process raster: /Users/chingkheirk/Library/Application Support/QGIS/QGIS3/profiles/default/python/plugins/SemiAutomaticClassificationPlugin/debug/raster.tif; calculation_datatype: <class 'numpy.float32'>
DEBUG_p1|2025-08-12T20:14:12.812|raster_vector|raster_info|252|/Users/chingkheirk/Library/Application Support/QGIS/QGIS3/profiles/default/python/plugins/SemiAutomaticClassificationPlugin/debug/raster.tif :[(304545.0, 30.0, 0.0, 4624365.0, 0.0, -30.0), 'metre', [8, 8], -99999.0, 1, [8, 8], [None, None], 'Float32', 'PROJCS["WGS 84 / UTM zone 33N",GEOGCS["WGS 84",DATUM["WGS_1984",SPHEROID["WGS 84",6378137,298.257223563,AUTHORITY["EPSG","7030"]],AUTHORITY["EPSG","6326"]],PRIMEM["Greenwich",0],UNIT["degree",0.0174532925199433,AUTHORITY["EPSG","9122"]],AUTHORITY["EPSG","4326"]],PROJECTION["Transverse_Mercator"],PARAMETER["latitude_of_origin",0],PARAMETER["central_meridian",15],PARAMETER["scale_factor",0.9996],PARAMETER["false_easting",500000],PARAMETER["false_northing",0],UNIT["metre",1,AUTHORITY["EPSG","9001"]],AXIS["Easting",EAST],AXIS["Northing",NORTH],AUTHORITY["EPSG","32633"]]']
DEBUG_p1|2025-08-12T20:14:12.812|processor|function_initiator|167|r_gt: (304545.0, 30.0, 0.0, 4624365.0, 0.0, -30.0)
DEBUG_p1|2025-08-12T20:14:12.812|processor|function_initiator|214|sections: [<remotior_sensus.core.processor.RasterSection object at 0x10d62d7f0>]
DEBUG_p1|2025-08-12T20:14:12.812|processor|function_initiator|425|_input_array.shape: (3, 8, 1)
DEBUG_p1|2025-08-12T20:14:12.812|processor|function_initiator|430|raster: /Users/chingkheirk/Library/Application Support/QGIS/QGIS3/profiles/default/python/plugins/SemiAutomaticClassificationPlugin/debug/raster.tif; band_number: 1
DEBUG_p1|2025-08-12T20:14:12.812|raster_vector|band_read_array_block|521|pixel_start_column: 0; pixel_start_row: 5; block_columns: 8; block_row: 3
DEBUG_p1|2025-08-12T20:14:12.813|processor|function_initiator|486|calculation_datatype: <class 'numpy.float32'>
DEBUG_p1|2025-08-12T20:14:12.813|processor|function_initiator|497|_input_array.shape: (3, 8, 1)
DEBUG_p1|2025-08-12T20:14:12.813|processor_functions|raster_unique_values_with_sum|1027|start
DEBUG_p1|2025-08-12T20:14:12.815|processor_functions|raster_unique_values_with_sum|1041|end
DEBUG_p1|2025-08-12T20:14:12.815|processor|function_initiator|564|output_raster_list: [None]
DEBUG_p1|2025-08-12T20:14:12.827|processor|function_initiator|683|end

DEBUG|2025-08-12T20:14:13.207|multiprocess_manager|_collect_results|579|start
DEBUG|2025-08-12T20:14:13.238|multiprocess_manager|_collect_results|708|end
DEBUG|2025-08-12T20:14:13.239|multiprocess_manager|run|353|end
DEBUG|2025-08-12T20:14:13.239|download_tools|download_file|98|url: https://www.python.org
DEBUG|2025-08-12T20:14:13.430|read_write_files|write_file|30|output_path: /private/var/folders/57/mx86lqjd62j_tvwlgs7pkvcm0000gn/T/semiautomatic_classification/remotior_sensus_0812eeh6dt3c/t20250812_201413239538_1952.html
DEBUG|2025-08-12T20:14:13.433|settings|test_dependencies|392|tests: <ul><li>NumPy: Success</li><li>SciPy: Success</li><li>Matplotlib: Success</li><li>GDAL: Success</li><li style="color:red">PyTorch: Fail</li><li>scikit-learn: Success</li><li>Remotior Sensus: Success</li><li>Multiprocess: Success</li><li>Internet connection: Success</li></ul>
INFO|2025-08-12T20:14:16.894|session|set|442|n_processes: 2; ram: 2048; temp.dir: /private/var/folders/57/mx86lqjd62j_tvwlgs7pkvcm0000gn/T/semiautomatic_classification/remotior_sensus_0812eeh6dt3c

**Comment by chinkgheirk** [comment]
The following step has solved the problem of "torch fail" 
Install PyTorch into the QGIS Python Environment:
Open your macOS Terminal.
Navigate to the directory containing the QGIS Python executable found in step 1. For example, if the executable is at /Applications/QGIS.app/Contents/MacOS/bin/python3, you would cd /Applications/QGIS.app/Contents/MacOS/bin/. 
Use the pip installer associated with this specific Python environment to install PyTorch. For example:
Code:
 ./python3 -m pip install torch torchvision torchaudio

**Comment by chinkgheirk** [comment]
The following steps solved my problem:

1. Identify the QGIS Python Environment:
QGIS typically uses its own bundled Python environment. Determine the path to this Python executable. One way to do this is by typing import sys; print(sys.executable) in the QGIS Python console.

2. Install PyTorch into the QGIS Python Environment:
Open your macOS Terminal.
Navigate to the directory containing the QGIS Python executable found in step 1. For example, if the executable is at /Applications/QGIS.app/Contents/MacOS/bin/python3, you would cd /Applications/QGIS.app/Contents/MacOS/bin/. 

Use the pip installer associated with this specific Python environment to install PyTorch. For example:
Code:   ./python3 -m pip install torch torchvision torchaudio

**Comment by alihassanijr** [comment]
I'm pretty sure you're posting this issue in the wrong place.

I have no idea what QGIS and ARCGIS are, and I don't see "natten" appear once in the issue.

Closing.
